src/Raven.Server/Documents/Replication/ReplicationDocumentSender.cs

```diff
@@ -647,6 +659,36 @@ private bool AddReplicationItemToBatch(ReplicationBatchItem item, OutgoingReplic
             return true;
         }
 
+        private bool ShouldSendAttachmentStream(AttachmentReplicationItem attachment)
+        {
+            if (MissingAttachmentsInLastBatch)
+            {
+                // we intentionally not trying to de-duplicate in this scenario
+                // we may have _sent_ the attachment already, but it was deleted at 
+                // destination, so we need to send it again
+                return true;
+            }
+                    
+            // RavenDB does de-duplication of attachments on storage, but not over the wire
+            // Here we implement the same idea, if (in the current connection), we already sent
+            // an attachment, we will skip sending it to the other side since we _know_ it is 
+            // already there.
+            if (_deduplicatedAttachmentHashes.Contains(attachment.Base64Hash))
+                return false; // we already sent it over during the current run
```

Comment:
Another thought that I had here..
Since you keep those in-mem what happened if you delete the attachment on the source and then recreate it.
I think that we should throw missing attachment and get it once more, but would like to have a test to verify this scenario,

Reply:
Yes, in this case, it will throw and then we'll be forced to send it again.
That is already part of the current behavior, and I was careful not to change it.